lib/elastic_apm/transport/connection.rb

```diff
@@ -4,203 +4,183 @@
 require 'concurrent'
 require 'zlib'
 
+require 'elastic_apm/transport/connection/proxy_pipe'
+require 'elastic_apm/transport/connection/state'
+
 module ElasticAPM
   module Transport
+    # rubocop:disable Metrics/ClassLength
     # @api private
-    class Connection # rubocop:disable Metrics/ClassLength
+    class Connection
       include Logging
 
-      class FailedToConnectError < InternalError; end
-
-      # @api private
-      # HTTP.rb calls #rewind the body stream which IO.pipes don't support
-      class ModdedIO < IO
-        def self.pipe(ext_enc = nil)
-          super(ext_enc).tap do |rw|
-            rw[0].define_singleton_method(:rewind) { nil }
-          end
-        end
-      end
-
       HEADERS = {
         'Content-Type' => 'application/x-ndjson',
         'Transfer-Encoding' => 'chunked'
       }.freeze
-      GZIP_HEADERS = HEADERS.merge('Content-Encoding' => 'gzip').freeze
+      GZIP_HEADERS = HEADERS.merge(
+        'Content-Encoding' => 'gzip'
+      ).freeze
 
-      # rubocop:disable Metrics/MethodLength
       def initialize(config, metadata)
         @config = config
-        @metadata = metadata.to_json
+        @metadata = JSON.fast_generate(metadata)
 
         @url = config.server_url + '/intake/v2/events'
-
-        headers =
-          (@config.http_compression? ? GZIP_HEADERS : HEADERS).dup
-
-        if (token = config.secret_token)
-          headers['Authorization'] = "Bearer #{token}"
-        end
-
-        @client = HTTP.headers(headers).persistent(@url)
-
-        configure_proxy
-        configure_ssl
+        @headers = build_headers
+        @client = build_client
+        @ssl_context = build_ssl_context
 
         @mutex = Mutex.new
+        @state = State.new
       end
-      # rubocop:enable Metrics/MethodLength
 
-      def configure_proxy
-        unless @config.proxy_address && @config.proxy_port
-          return
-        end
-
-        @client = @client.via(
-          @config.proxy_address,
-          @config.proxy_port,
-          @config.proxy_username,
-          @config.proxy_password,
-          @config.proxy_headers
-        )
-      end
-
-      def configure_ssl
-        return unless @config.use_ssl? && @config.server_ca_cert
-
-        @ssl_context = OpenSSL::SSL::SSLContext.new.tap do |context|
-          context.ca_file = @config.server_ca_cert
-        end
-      end
+      attr_reader :state
 
+      # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
       def write(str)
         return if @config.disable_send
 
-        connect_unless_connected
-
-        @mutex.synchronize { append(str) }
+        connect
+        append(str)
 
-        return unless @bytes_sent >= @config.api_request_size
+        return true unless @wr.bytes_sent >= @config.api_request_size
 
-        flush
-      rescue FailedToConnectError => e
-        error "Couldn't establish connection to APM Server:\n%p", e
-        flush
+        flush(:api_request_size)
```

Comment:
At this point the state `disconnected` could be returned, and you'd still return `true` for `write` although nothing was written.

Reply:
I might not have had enough coffee yet. Can you describe it in more detail?

Reply:
I might be overlooking something here, but `flush` simply returns if in state `disconnected`. What happens if in between `connect` and `flush` a disconnect happens? (Or can this not happen?)

Reply:
Hopefully this will make sure to not let it hang: https://github.com/elastic/apm-agent-ruby/blob/243cc91603a8ef768a2dbd2c6ba1038160e2be9e/lib/elastic_apm/transport/connection.rb#L168-L175